### ISSUE TITLE
Examples for common usages

### ISSUE THREAD

**Issue by sgtcoolguy** [jenkinsci/jira-steps-plugin#2]
I'd love to make use of this plugin in our new pipeline jobs since the official JIRA plugin doesn't support some major operations in pipelines (like issue migration).

Specifically, one common thing we tend to do is to gather the list of JIRA tickets affected by the commits, update the JIRA tickets using the JIRA plugin (which adds comments and backers to the build), then create a new version in JIRA if it doesn't already exist, then "migrate" all the affected tickets to add that new version as a fixVersion.

I see that I can use the combination of both this plugin and the JIRA plugin to get the list of affected issue keys and update the issues(JIRA plugin). Is there any way to determine if a version exists by using the display name (and not the id, which I wouldn't know) with any steps in this plugin? Or maybe a step to createOrGetVersion by name?

Once I create a version, does that turn an updated Version object with the new id in it? Otherwise how can I set the fixVersion on the issues to point at that new version?

I guess there's a disconnect for me on how I'd handle if a version existed (and I only knew the name, not the id) and still be able to update the issues affected; or even if creating a version gives me back the object or id I need to update the issues' fixVersion.

**Comment by nrayapati** [comment]
Hello Welcome to JIRA Steps. 

To update the issue's fix version, id is not mandatory, it can be just a name, but it doesn't complain if we send id as well. 

Just in case if we have to know id  of the version/component there is no straight step(s) yet, but we could use the `jiraGetProject` to know all the versions/components. 

### Query Project to Get Versions
For example querying the project TEST and printing the versions:
```groovy
node {
  stage('JIRA') {
    def projectResults = jiraGetProject idOrKey: "TEST"
    echo projectResults.data.versions.toString()
 }
}
```

Output:
```groovy
[Pipeline] jiraGetProject
JIRA: Site - LOCAL - Querying Project with idOrKey:TEST
Successful. Code: 200
[Pipeline] echo
[Version(id=10000, name=test-version-1.0, description=test-version-1.0, archived=false, released=false, startDate=null, releaseDate=null, userStartDate=null, userReleaseDate=null, project=null, projectId=10000), Version(id=10100, name=test-version-1.1, description=null, archived=false, released=false, startDate=null, releaseDate=null, userStartDate=null, userReleaseDate=null, project=null, projectId=10000)]
```

### Updating Issue's Fix version:

As I said above, version id is not mandatory field, following is the example that I tried to mimic the situation that you are handling.  

Query issues by jql `jiraJqlSearch` and loop though the result (basic issue output) and query back the original issue to know more about that issue and update these issues accordingly, this case with the new/existing version. 

Certainly there are few things to improve, the documentation and minor tweaks in the code. Which I am talking look at in my free time. :)

```
node {
  stage('JIRA') {
    def searchResults = jiraJqlSearch jql: "project = TEST AND issuekey = 'TEST-17'"
    def issues = searchResults.data.issues
    for (i = 0; i <issues.size(); i++) {
      def result = jiraGetIssue idOrKey: issues[i].key
      def issue = result.data

      def fixVersion = [name: "existing-fix-version-1.0"]

      // or if we have to create new version 
      // fixVersion = jiraNewVersion version: [name: "existing-fix-version-1.0"]
      // We could use the above variable directly, but actually JIRA accepts only string while updating issue, but I designed class to be integer. 
      // Which is causing some problem there.
      // fixVersion = [name: fixVersion.name]
    
      def testIssue = [fields: [ project: issue.fields.project,
                                 summary: issue.fields.summary,
                                 description: issue.fields.description,
                                 issuetype: issue.fields.issuetype,
                                 fixVersions: [[name: "test-version-1.0"]]]]
      response = jiraEditIssue idOrKey: issues[i].key, issue: testIssue
    }
  }
}
```

Note: The above example could be simplified a bit while updating the issue it can only be just fixversions field, but due the input validation that I have added which is expecting all the remaining fields too. 

More notes: https://github.com/jenkinsci/jira-steps-plugin/projects/1

**Comment by nrayapati** [comment]
@sgtcoolguy Have fixed few issues that I talked above with #3 and added a bit of documentation[1].

Hope examples here was a bit useful and answered your question. Thanks.

[1] https://jenkinsci.github.io/jira-steps-plugin/common_usages.html

General Note: every step has some common input fields (`site` and `failOnError`) and also common output fields (`data`, `error`, `code`, `successful`). as seen above `data` should have response object, which can be used to further invoke other steps by passing this as an input.

**Comment by nrayapati** [comment]
@sgtcoolguy Hope above examples helped you a bit. For now closing this issue. just in case if you still have some questions around please don't hesitate to reopen or create a new issue. Thanks.
